src/Sulu/Bundle/SnippetBundle/Content/SnippetDataProvider.php

```diff
@@ -220,6 +240,37 @@ function(
         );
     }
 
+    /**
+     * @return array<int, array<string, string>>
+     */
+    private function getTypes(): array
+    {
+        $types = [];
+        if ($this->tokenStorage && null !== $this->tokenStorage->getToken()) {
+            // the user is required for the metadata locale because of the translated template titles
+            $user = $this->tokenStorage->getToken()->getUser();
+
+            // user is "anon." if the user is not logged in
+            // this happens on the initial admin request (/admin) if the AudienceTargeting is deactivated
+            if (!$user instanceof UserInterface) {
+                return $types;
+            }
+
+            try {
```

Comment:
The `try/catch` is necessary because in some tests like https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/AudienceTargetingBundle/Tests/Functional/Controller/AdminControllerTest.php#L32 there is no `SnippetDataProvider` registered, but I don't really have an idea why the providers are not registered.
Do you have any idea? @danrot @wachterjohannes

Reply:
@Prokyonn is the snippet bundle registered in the test app?

Reply:
Yes, the `Kernel` extends the `SuluTestKernel` and this kernel registers the `SnippetBundle`

Reply:
Do you have a link to the failing test?

Reply:
Not in the CI, but the `testRulesConfig` of the audiencetargeting bundle fails https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/AudienceTargetingBundle/Tests/Functional/Controller/AdminControllerTest.php#L32

Reply:
@alexander-schranz found a solution. The error happens because in those bundles no snippet templates exist. Now the question is if we should add a default snippet template in each test bundle or if it is a valid case, if there is really no snippet template available. WDYT @danrot

Reply:
I would say the system should not break just because there are no snippets defined... The thing is that the [`StructureFormMetadataLoader`](https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/AdminBundle/Metadata/FormMetadata/StructureFormMetadataLoader.php#L96) has to return `null` here, because the mechanism relies on `MetadataLoader`s to not return anything, in case there is nothing... And the `StructureMetadataFactory` does not seem to return `snippet` as a structure type if no snippets are given...

Maybe (not sure, have not tested it) it would be enough to instead of [implicitly creating the array](https://github.com/sulu/sulu/blob/0da80a9562d5249dd629a45d0fe18281fdf85f10/src/Sulu/Bundle/AdminBundle/Metadata/FormMetadata/StructureFormMetadataLoader.php#L129) make that explicit, i.e. create the array like this after the foreach in line 123:

```php
$structuresMetadataByTypes[$structureType] = [];
```

Then snippets should be returned as an option even if no snippets are available. Not sure if that will create other problems though :see_no_evil: But I guess that would be the proper fix to this problem.